RavenDB-14634 avoid blocking record change

karmeli87 opened the pull request:
- don't create TCP connection for internal nodes under lock
- don't dispose Index under lock
- don't dispose ETL under lock
- don't wait for conflicts resolution under lock

karmeli87 committed:
RavenDB-14634 avoid blocking record change

- don't create TCP connection for internal nodes under lock
- don't dispose Index under lock
- don't dispose ETL under lock
- don't wait for conflicts resolution under lock

karmeli87 committed:
RavenDB-14634 add debug logs

karmeli87 committed:
RavenDB-14634 rebase and cleanup

ayende commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -13,21 +13,22 @@
-        private EtlProcess[] _processes = new EtlProcess[0];
-        private readonly HashSet<string> _uniqueConfigurationNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase); // read and modified under a lock.
+        private readonly ConcurrentSet<EtlProcess> _processes = new ConcurrentSet<EtlProcess>();
+        private readonly ConcurrentSet<string> _uniqueConfigurationNames = new ConcurrentSet<string>(StringComparer.OrdinalIgnoreCase); // read and modified under a lock.
```
Is the comment here still accurate?

ayende commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -113,8 +109,10 @@ private void OnProcessRemoved(EtlProcess process)
                 if (newSqlDestinations != null && newSqlDestinations.Count > 0)
                     newProcesses.AddRange(GetRelevantProcesses<SqlEtlConfiguration, SqlConnectionString>(newSqlDestinations, ensureUniqueConfigurationNames));
 
-                processes.AddRange(newProcesses);
-                _processes = processes.ToArray();
```
Why not keep the separate instance model?

ayende commented on src/Raven.Server/Documents/Indexes/Index.cs:
```diff
@@ -277,6 +277,23 @@ protected virtual void DisposeIndex()
 
                 exceptionAggregator.Execute(() => { _indexingProcessCancellationTokenSource?.Dispose(); });
 
+                if (delete)
```
Who calls this with `true` ?

ayende commented on src/Raven.Server/Documents/Indexes/IndexStore.cs:
```diff
@@ -949,7 +1000,7 @@ private void DeleteIndexInternal(Index index, bool raiseNotification = true)
             try
             {
-                index.Dispose();
+                index.Dispose(delete: true);
```
Why did you move the delete to here?

ayende commented on src/Raven.Server/Documents/Replication/ReplicationLoader.cs:
```diff
@@ -960,8 +970,21 @@ private void HandleInternalReplication(DatabaseRecord newRecord, List<IDisposabl
                     Database = Database.Name
+                }).ToList();
+
+                Task.Run(() =>
```
Cancellation token?

ayende commented on src/Sparrow/Utils/ThreadingHelper.cs:
```diff
@@ -0,0 +1,21 @@
+                initialValue = location;
+                if (initialValue >= newValue)
+                    return false;
+            } while (Interlocked.CompareExchange(ref location, newValue, initialValue) != initialValue);
```
Better to set the `initialvalue` to the result of `CompareExchange` here.

ppekrol commented:
@karmeli87 Is is ready for merge? Can you rebase?

karmeli87 committed:
RavenDB-14634 fix tests

karmeli87 committed:
RavenDB-14634 avoid blocking record change

- don't create TCP connection for internal nodes under lock
- don't dispose Index under lock
- don't dispose ETL under lock
- don't wait for conflicts resolution under lock

karmeli87 committed:
RavenDB-14634 fix add node EP

- use UpdateTopologyCommand for adding node to database
- fix ShouldSetClientConfigEtag behaviour
- refresh topology/client-config only if needed

karmeli87 committed:
RavenDB-14634 propogate the real error from the leader

karmeli87 committed:
RavenDB-14634 enhance Rehab logic

- Give grace time based on the last topology node modification
- Don't move to rehab if the database was recently added to the landlord

karmeli87 committed:
RavenDB-14634 make sure changes are connected

fixes a race in LazyLoadResultShouldBeUpToDateEvenIfAggressiveCacheIsEnabled

karmeli87 commented:
Closed in favor of a slimmer version: https://github.com/ravendb/ravendb/pull/12517
